### ISSUE TITLE
Compatibility with other Postgres libraries

### ISSUE THREAD

**Issue by kdevan** [tpetry/laravel-postgresql-enhanced#42]
Hello! I use Postgres heavily and am very excited to see that you started this project. 

Right now I have `umbrellio/laravel-pg-extensions` implemented and have used its "Custom Extensions" method for adding other new functionality I needed. On top of that library I have `belamov/postgres-range`, `staudenmeir/laravel-adjacency-list`, and with it `staudenmeir/laravel-cte`. 

Most of the custom functionality I've added is already handled by your library, so I would just be able to remove that on my end. You also support ranges so I'm not worried about that part. I do use Laravel Adjacency List heavily so I'm wondering how it would work with both libraries implementing CTE, and then compatibility with other Postgres libraries in general.

I think the main issue I run into when I'm trying to get these libraries to work together is when there's conflicts with the PostgresConnection. And then some way for multiple libraries to extend Blueprint and Schema Grammar which is what the [Custom Extension method](https://github.com/umbrellio/laravel-pg-extensions#custom-extensions) allows for. 

So far I've used the `umbrellio/laravel-pg-extensions` implementation of PostgresConnection and extended that. Then when there's conflicts, which I ran into with `belamov/postgres-range` and `staudenmeir/laravel-cte`, I just implement them manually using the Custom Extension method so that they do not load their versions of the PostgresConnection, and prevent them from being automatically discovered by Laravel using composer.json:
```json
"extra": {
    "laravel": {
        "dont-discover": [
          "belamov/postgres-range",
          "staudenmeir/laravel-cte"
        ]
    }
}
```
Switching to your library I won't have that Custom Extension method anymore, which means I don't have a clear way to deal with conflicts, so I know that will be the main challenge in transitioning here. 

Of course this is out of scope of this repository but I thought I'd ask just in case you have some insight here. Either way, I do plan to try this transition in just the next few days, so if you're not sure I'll go ahead and just update this issue with anything that I find, maybe help others in the future.

I am so excited about this library, thank you so much for starting this!

**Comment by tpetry** [comment]
Yeah that the database part is not extensible by multiple libraries is a problem in Laravel, but I didn't like the solution of `umbrellio/laravel-pg-extensions` because for any conflict you need to know exactly what to change. The CTE implementation is a good example as a lot of different other code parts need to be changed to support it.

So it is basically that you can only use one library to extend the database stuff. I designed my CTE implementation to be kind of similar to `staudenmeir/laravel-cte`. The basic syntax is the same, but the extended syntax for e.g. recursive queries or materialization is different.

So I don't know the problems you will face, but forking the other packages to change syntax a bit should work. You can share your findings about incompatibilities but I don't promise to work on them. You are entering a part of Laravel that is very fickle and it was kind of never expected that multiple plugins extend the db functionality.

**Comment by kdevan** [comment]
Ah well thank you, this is very helpful context going into this.

> You can share your findings about incompatibilities but I don't promise to work on them.

Of course, I'll be happy to share information here just on the chance it helps anyone else :)

**Comment by tpetry** [comment]
I had the plan to make this extensibility when I started the library. But the more features I added, I understood that this idea is impossible. For adding support X maybe also feature Y needs to be changed. Its just not possible to let multiple packages extend the db driver without having so many conflicts or package Y needing to change its implementation because package X wants to add something.

**Comment by tpetry** [comment]
Adding casts like `belamov/postgres-range` to fully replace it should be pretty easy.

But `umbrellio/laravel-ltree` provides a lot of stuff. And I don't know whether I want to support all of that. What are you using from this package?

**Comment by tpetry** [comment]
I've pinned the issue as this problem may occur in the future again for other developers.
